### PR TITLE
Move and merge verifier of Set/GetDimensionSizeOp

### PR DESCRIPTION
**Pull request by zhouxin913**
SetDimensionSizeOp is very special, compared to other ops when involving bounded dynamism: it is possible that the output have bounds while the inputs has no bounds at all. Thus we can't use this util function to generate the encoding at the end of shape function: 
```
Attribute boundsToEncoding(Attribute prototype, ArrayRef<int64_t> bounds)
```
Because the `prototype`, which comes from `input.getEncoding()` maybe null. So in this PR it create bounds attr from dialect directly. 

Change `inferReturnTypes` to `inferReturnTypeComponents` and remove the TODO in comment as well.

### CODE REVIEW

File: stablehlo/dialect/TypeInference.cpp [hunk 1]
```diff
@@ -2689,6 +2703,55 @@ LogicalResult inferSendOp(Dialect* dialect, std::optional<Location> location,
   return success();
 }
 
+LogicalResult inferSetDimensionSizeOp(
+    Dialect* dialect, std::optional<Location> location, Type operandType,
+    Value size, int64_t dim,
+    SmallVectorImpl<ShapedTypeComponents>& inferredReturnShapes) {
+  if (auto sizeType = size.getType().dyn_cast<RankedTensorType>()) {
+    if (sizeType && sizeType.getRank() != 0)
```
Comment (sdasgup3) [inline stablehlo/dialect/TypeInference.cpp:2711]: no need to check `sizeType` as it has already been checked above.

Comment (sdasgup3) [reply stablehlo/dialect/TypeInference.cpp:2711]: Or maybe
```
auto sizeType = size.getType().dyn_cast<RankedTensorType>();
if (sizeType && sizeType.getRank() != 0)
  emitOptionalError(location, "size operand should be of rank-0");
```

**Review by sdasgup3 — APPROVED** [review]

File: stablehlo/dialect/TypeInference.cpp (hunk 1 above)
Comment (zhouxin913) [reply stablehlo/dialect/TypeInference.cpp:2711]: Sure updated as your suggested.

File: stablehlo/dialect/TypeInference.h [hunk 2]
```diff
@@ -231,8 +231,8 @@ LogicalResult inferGatherOp(
     SmallVectorImpl<ShapedTypeComponents>& inferredReturnShapes);
 
 LogicalResult inferGetDimensionSizeOp(
-    MLIRContext* context, std::optional<Location> location,
-    SmallVectorImpl<Type>& inferredReturnTypes);
+    MLIRContext* context, std::optional<Location> location, Type operandType,
+    int64_t dim, SmallVectorImpl<ShapedTypeComponents>& inferredReturnShapes);
```
Comment (burmako) [inline stablehlo/dialect/TypeInference.h:235]: `dimension`, according to how this is called in the ODS? (Same below).

File: stablehlo/dialect/TypeInference.h [hunk 3]
```diff
@@ -231,8 +231,8 @@ LogicalResult inferGatherOp(
     SmallVectorImpl<ShapedTypeComponents>& inferredReturnShapes);
 
 LogicalResult inferGetDimensionSizeOp(
-    MLIRContext* context, std::optional<Location> location,
-    SmallVectorImpl<Type>& inferredReturnTypes);
+    MLIRContext* context, std::optional<Location> location, Type operandType,
```
Comment (burmako) [inline stablehlo/dialect/TypeInference.h:234]: Now that we're passing in `operandType`, we can drop the `context` parameter and obtain it from the type - `operandType.getContext()`.

File: stablehlo/dialect/TypeInference.cpp [hunk 4]
```diff
@@ -2689,6 +2703,54 @@ LogicalResult inferSendOp(Dialect* dialect, std::optional<Location> location,
   return success();
 }
 
+LogicalResult inferSetDimensionSizeOp(
+    Dialect* dialect, std::optional<Location> location, Type operandType,
+    Value size, int64_t dim,
+    SmallVectorImpl<ShapedTypeComponents>& inferredReturnShapes) {
+  auto sizeType = size.getType().dyn_cast<RankedTensorType>();
+  if (sizeType && sizeType.getRank() != 0)
+    emitOptionalError(location, "size operand should be of rank-0");
```
Comment (burmako) [inline stablehlo/dialect/TypeInference.cpp:2712]: `return emitOptionalError(...)`?

**Review by burmako — CHANGES_REQUESTED** [review]

File: stablehlo/dialect/TypeInference.cpp (hunk 4 above)
Comment (zhouxin913) [reply stablehlo/dialect/TypeInference.cpp:2712]: Good catch! Involved when addressing comments.

File: stablehlo/dialect/TypeInference.h (hunk 3 above)
Comment (zhouxin913) [reply stablehlo/dialect/TypeInference.h:234]: Sure

File: stablehlo/dialect/TypeInference.h (hunk 2 above)
Comment (zhouxin913) [reply stablehlo/dialect/TypeInference.h:235]: Sure.

**Review by burmako — APPROVED** [review]

**Pull request merged**